<a href="https://colab.research.google.com/github/AnnaBo1/python_for_DA/blob/main/%226_2_%D0%A1%D1%82%D0%B0%D1%82%D0%B8%D1%81%D1%82%D0%B8%D1%87%D0%BD%D1%96_%D1%82%D0%B5%D1%81%D1%82%D0%B8_ipynb%22.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Домашнє завдання 6: Статистичні тести

0. Зчитайте дані з `data.csv` та запишіть значення колонки data у форматі np.ndarray в змінну data. Ми будемо далі працювати з цією змінною.

In [22]:
import pandas as pd
import numpy as np

# Зчитуємо дані з CSV
df = pd.read_csv('data.csv')

# Отримуємо колонку 'data' як np.ndarray
data = df['data'].values

print(data)

[17.49945292 19.66239904  7.18282343 29.84162485  9.23938649 14.94951581
 23.0172885  12.52827148 13.65228669 14.54595431 23.30872427 33.75324808
 20.24923636 13.29244733 23.23434992 16.4230418  19.88521702 27.05000732
 15.5127743  20.05415151 14.73135264 19.06139498 21.53942271 14.06732571
 17.9670682  18.58289581 16.17406993 12.87432628 11.47269664 19.07902883
 18.38565824 33.38820073  5.39139454 20.67635903 22.22266722 28.15780318
 23.01114324 14.93471778 20.00005857 23.25411543 18.11895082 24.62607043
  8.79145607 30.387108   28.80606806 17.98593597 23.66804468 20.28782355
 15.02518827 20.52626131 26.00219532 17.71344489 17.74598346 19.55317542
 22.60097798 27.67027538 16.19192417 23.05037746 21.29669604  8.84832568
 17.48410111 19.20602661 19.76257856 21.9560206   7.75806171 20.27753314
 15.93394654 11.36336584 23.14577858 24.41167746 16.08049839 25.05473769
 17.71090111 20.39893405 13.40756632 29.50692234  4.04330326 19.45128426
 24.17071763  7.79920072 18.86318441 19.53668801 24

Запустіть код нижче. Для коректної робити всіх подальших методів дані мають бути у вигляді одновимірного вектора, тобто мати розмір (100,).

In [23]:
data.shape

(100,)

1. Зчитані дані - це сума покупок в доларах наших покупців на сайті протягом липня 2022 року.

До цього місяця, ми мали продажі в середньому на чек $20. Чи вийшло у нас статистично значущо **підвищити** середній чек за останній місяць?

Оберіть статистичний тест (z-test чи t-test). Виведіть p-значення статистичного тесту та зробіть висновок, чи ми маємо зараз статистично значущу різницю у розмірі середнього чеку.

**Підказка**: тут тест буде одновибірковий.

In [24]:
import numpy as np
from scipy import stats  # імпорт для t-тесту

In [25]:
mean = 20  # Попереднє середнє значення чеку

#Використання t-тесту
t_stat, p_value = stats.ttest_1samp(data, mean)

print(t_stat)
print(p_value)

-0.9953477389335053
0.3219937839011988


T-статистика: Показує, наскільки далеко від попереднього середнього значення розташовується новий середній чек.

P-значення: Вказує на ймовірність отримати такі результати (або ще більш екстремальні) під умовою, що нульова гіпотеза вірна (тобто, середній чек не змінився).

t-статистика: Значення t-статистики близьке до нуля, що вказує на те, що середній чек не відрізняється значно від порівняльного значення ($20).

p-значення 0.322 є значно більшим за стандартний рівень значущості 0.05.
Це свідчить про те, що ми не можемо відкинути нульову гіпотезу.

Середній чек не змінився статистично значущо у порівнянні з попереднім середнім чеком у $20. Це означає, що за останній місяць не було статистично значущого підвищення середнього чека на сайті

2. Пройшов час, і ми зібрали дані про продажі за серпень 2022 року, які зберігли у файл `data_2.csv`. Продажів було дещо більше, але чи змінився стат. значущо середній чек порівняно з липнем 2022? Перевірте це використовуючи ztest.

  Виведіть p-значення статистичного тесту та зробіть висновок, чи ми маємо зараз статистично значущу різницю у розмірі середнього чека між двома місяцями.

  **Підказка**: треба провести двовибірковий ztest з Н0 - що середні значення в липні та серпні 2022 року - однакові. Для кращого розуміння ситуації, можливо вам буде цікаво переглянути, які середні показники чеку були в кожен з місяців. Порівняйте свій висновок з простого спостереження середніх значень і висновок статистичного тесту. Чи вони збігаються?

In [26]:
from statsmodels.stats.weightstats import ztest  # Імпорт для z-тесту

In [27]:
# Зчитуємо дані з CSV
data_july = pd.read_csv('data.csv')      # Липень
data_august = pd.read_csv('data_2.csv')  # Серпень

In [28]:
# Отримуємо колонки з даними про чеки
july_data = data_july['data'].values
august_data = data_august['data'].values #.values-конвертує дані у формат NumPy масиву (np.ndarray), що дозволяє виконувати математичні операції.

In [29]:
# Обчислюємо середні значення та стандартні відхилення
mean_july = np.mean(july_data)
mean_august = np.mean(august_data)

In [30]:
# Вибіркове стандартне відхилення
std_july = np.std(july_data, ddof=1)  #обчислює вибіркове стандартне відхилення для липневих даних. Параметр ddof=1 означає, що ми використовуємо корекцію на кількість ступенів свободи (n-1).
std_august = np.std(august_data, ddof=1)

In [31]:
# Виводимо середні значення для порівняння
print(mean_july)
print(mean_august)

19.37755319670441
19.37322276512519


In [35]:
# Проведення двовибіркового z-тесту
z_stat, p_value = ztest(july_data, august_data)  # виконує двовибірковий z-тест, порівнюючи два набори даних (липень та серпень) і повертає z-статистику (z_stat) і p-значення (p_value).

In [36]:
print(z_stat)
print(p_value)

0.005246673456808504
0.9958137794593356


In [37]:
# Висновок
if p_value < 0.05:  #перевіряє, чи p-значення менше 0.05 (зазначений рівень значущості)
    print("Середній чек статистично значущо змінився між липнем та серпнем.")
else:
    print("Середній чек не змінився статистично значущо між липнем та серпнем.")

Середній чек не змінився статистично значущо між липнем та серпнем.


При порівнянні середніх значень ми бачимо, що середній чек у липні (19.38) дещо перевищує середній чек у серпні (19.37). Різниця між цими значеннями є дуже незначною, і на перший погляд може здатися, що середній чек у липні вищий за серпневий.

P-значення(p_value): Значення 0.9958 вказує на те, що ймовірність отримати такі або більш екстремальні результати, якщо нульова гіпотеза (H0: середні значення в липні та серпні однакові) вірна, є дуже високою. Зазвичай, якщо p-значення менше 0.05, ми відкидаємо нульову гіпотезу, однак в даному випадку p-значення значно більше 0.05.

На основі простого спостереження середніх значень можна стверджувати, що середній чек у липні трохи вищий, ніж у серпні. Однак, результати статистичного тесту (z-test) свідчать про те, що ця різниця не є статистично значущою.

Отже, нульова гіпотеза не відхиляється, і ми можемо зробити висновок, що немає достатніх підстав стверджувати, що середній чек у серпні відрізняється від середнього чеку у липні.